modules/javafx.controls/src/test/java/test/javafx/scene/control/TestListenerHelper.java

```diff
@@ -0,0 +1,465 @@
+        SimpleStringProperty p = new SimpleStringProperty();
+        AtomicInteger ct = new AtomicInteger();
+
+        h.addChangeListener(p, true, (cur) -> ct.incrementAndGet());
+
+        p.set("1");
+        assertEquals(2, ct.get());
+
+        h.disconnect();
+
+        p.set("2");
+        assertEquals(2, ct.get());
+    }
+
+    // invalidation listeners
+
+    @Test
+    public void testInvalidationListener_MultipleProperties() {
+        ListenerHelper h = new ListenerHelper();
+        SimpleStringProperty p1 = new SimpleStringProperty();
+        SimpleStringProperty p2 = new SimpleStringProperty();
+        AtomicInteger ct = new AtomicInteger();
+
+        h.addInvalidationListener(() -> ct.incrementAndGet(), p1, p2);
+
+        p1.set("1");
+        p2.set("2");
+        assertEquals(2, ct.get());
+
+        h.disconnect();
+
+        p1.set("3");
+        p2.set("4");
+        assertEquals(2, ct.get());
+    }
+
+    @Test
+    public void testInvalidationListener_MultipleProperties_FireImmediately() {
+        ListenerHelper h = new ListenerHelper();
+        SimpleStringProperty p1 = new SimpleStringProperty();
+        SimpleStringProperty p2 = new SimpleStringProperty();
+        AtomicInteger ct = new AtomicInteger();
+
+        h.addInvalidationListener(() -> ct.incrementAndGet(), true, p1, p2);
+
+        p1.set("1");
+        p2.set("2");
+        assertEquals(3, ct.get());
+
+        h.disconnect();
+
+        p1.set("3");
+        p2.set("4");
+        assertEquals(3, ct.get());
+    }
+
+    @Test
+    public void testInvalidationListener_Plain() {
+        ListenerHelper h = new ListenerHelper();
+        SimpleStringProperty p = new SimpleStringProperty();
+        AtomicInteger ct = new AtomicInteger();
+
+        h.addInvalidationListener(p, (x) -> ct.incrementAndGet());
+
+        p.set("1");
+        assertEquals(1, ct.get());
+
+        h.disconnect();
+
+        p.set("2");
+        assertEquals(1, ct.get());
+    }
+
+    @Test
+    public void testInvalidationListener_Plain_FireImmediately() {
+        ListenerHelper h = new ListenerHelper();
+        SimpleStringProperty p = new SimpleStringProperty();
+        AtomicInteger ct = new AtomicInteger();
+
+        h.addInvalidationListener(p, true, (x) -> ct.incrementAndGet());
+
+        p.set("1");
+        assertEquals(2, ct.get());
+
+        h.disconnect();
+
+        p.set("2");
+        assertEquals(2, ct.get());
+    }
+
+    @Test
+    public void testInvalidationListener_Callback() {
+        ListenerHelper h = new ListenerHelper();
+        SimpleStringProperty p = new SimpleStringProperty();
+        AtomicInteger ct = new AtomicInteger();
+
+        h.addInvalidationListener(p, (cur) -> ct.incrementAndGet());
+
+        p.set("1");
+        assertEquals(1, ct.get());
+
+        h.disconnect();
+
+        p.set("2");
+        assertEquals(1, ct.get());
+    }
+
+    @Test
+    public void testInvalidationListener_Callback_FireImmediately() {
```

Comment:
Minor: redundant (tests the same method signature as `testInvalidationListener_Plain_FireImmediately`.

Reply:
another bug in tests.
+1